internal/handlers/common/aggregations/aggregations.go

```diff
@@ -24,6 +24,31 @@ import (
 
 // Stage is a common interface for all aggregation stages.
 type Stage interface {
+	// FirstStage fetches document iterator, it is only used if this is
+	// the first stage of the pipeline.
+	// The returned document iterator may be originated from querying database or
+	// from in-memory value iterator.
+	// This allows first stage of the pipeline to decide how to create initial document iterator.
+	FirstStage(ctx context.Context, closer *iterator.MultiCloser) (types.DocumentsIterator, error)
```

Comment:
I don't think we need that method at all. First of all, it looks suspiciously like a constructor; it is almost as if the `io.Reader` interface had `NewReader` method. Second, the implication there is that `Stage` has access to the data source, but then it is not clear why `Process` can't use it.

One option could be to fit everything into the existing `Stage` interface with only the `Process` method. `$collStats` implementation, for example, could ignore incoming `iter` and use the data source instead. Then the code that arranges the pipeline (a sequence of stages in this case) could check that the order is correct.

Another option is to say that pipelines are not just a sequence of stages but a combination of zero or one Producer stage and zero or more Processor stages. That way, `$collStats` would implement only the `Producer` interface, `$count` would implement only the Processor interface, and `$facet` would implement both. Then the compiler itself could check the order.

Reply:
Both are definitely better options. First option being simpler. The second option seems efficient :+1:

Reply:
I think using `ProducerStage` and `ProcessorStage` works quite well, thanks for the suggestion. Updated implementation a bit :hugs: